### ISSUE TITLE
default VOC validation incorrect for batch_size > 1

### ISSUE THREAD

**Issue by HaydenFaulkner** [dmlc/gluon-cv#872]
This issue appears related to #536 and #537. When validating with the voc metric on voc data, the input into the metric's `update()` function are lists of `len() = num_gpus` which contain the `mxnet.NDArray`'s of size `(B, N, ...)`. However the `update()` function expects the input to have the first dimension as the batch size, not the number of gpus. This caused a different evaluation results for different batch sizes. If the `batch_size / num_gpus = 1` the desired results are given, otherwise if the batch_size > 1 (or in fact `gt_labels[0].shape[0] != 1`) there is incorrect overwriting of `gt_difficults` with `None`s:

For example when `n_gpu=1, batch_size=2`:
```python
if isinstance(gt_labels, list):  # will be True
    if len(gt_difficults) != len(gt_labels) * gt_labels[0].shape[0]:  # will be True --> 1 != 1 * 2
         gt_difficults = [None] * len(gt_labels) * gt_labels[0].shape[0]  # now gt_difficults = [None, None]
```
Two ways to fix this could be:
1. To ensure that the input to the `update()` function meets the function requirements and is of shape `(B, N, ...)` which can be easily done by applying the `as_numpy()` function seen in the voc metric to all inputs (in the training scripts):
```python
eval_metric.update(as_numpy(det_bboxes), as_numpy(det_ids), as_numpy(det_scores), as_numpy(gt_bboxes), as_numpy(gt_ids), as_numpy(gt_difficults))
```

2. Fix it in the `update()` function itself, however I'm not sure what the purpose of the offending code is, so unsure if a fix would break its purpose? Simply removing the 3 lines fix the issue:
https://github.com/dmlc/gluon-cv/blob/71a0d545ba2141c3fd0a54708d088bc8c8029dfb/gluoncv/utils/metrics/voc_detection.py#L108-L110
Or maybe something like below would be more appropriate?
```python
if isinstance(gt_labels, list):
    if len(gt_difficults) * gt_difficults[0].shape[0] != len(gt_labels) * gt_labels[0].shape[0]:
        gt_difficults = [None] * len(gt_labels) * gt_labels[0].shape[0]
```

**Comment by zhreshold** [comment]
actually it's the faulty line here to be fixed: https://github.com/dmlc/gluon-cv/blob/71a0d545ba2141c3fd0a54708d088bc8c8029dfb/gluoncv/utils/metrics/voc_detection.py#L106

A fix could be:
```python
if gt_difficults is None:
    if isinstance(gt_labels, list):
        num_gt = sum([x.shape[0] for x in gt_labels])
    else:
        num_gt = gt_labels.shape[0]
    gt_difficults = [None] * num_gt
```

**Comment by HaydenFaulkner** [comment]
But using the `VOCDetection` dataset `gt_difficults` is not `None` but rather a `list` of `mxnet.NDArray`'s so that won't resolve my issue. The incorrect mAP scores are attained because when `gt_difficults` are passed in they are (incorrectly?) replaced with `None`s in (since `gt_labels` is also a `list`):
https://github.com/dmlc/gluon-cv/blob/71a0d545ba2141c3fd0a54708d088bc8c8029dfb/gluoncv/utils/metrics/voc_detection.py#L108-L110
The removal of the actual `gt_difficults` values leads to a mAP of .76, compared to .83 when `gt_difficults` is maintained.

**Comment by zhreshold** [comment]
I see, could you contribute a PR where the faulty if line 
```if len(gt_difficults) != len(gt_labels) * gt_labels[0].shape[0]:```
will be fixed?
